langfuse/src/openai.ts

```diff
@@ -0,0 +1,256 @@
+import OpenAI from "openai";
+import { Langfuse, LangfuseTraceClient } from "./langfuse";
+
+
+const client = new Langfuse();
+interface InputArgsData {
+    model: string
+    input: Record<string, any>
+    modelParams: Record<string, any>
+}
+
+const parseInputArgs = (args: Record<string, any>): InputArgsData => {
+    let params: Record<string, any> = {}
+    params = {
+        frequency_penalty: args.frequency_penalty,
+        logit_bias: args.logit_bias,
+        logprobs: args.logprobs,
+        max_tokens: args.max_tokens,
+        n: args.n,
+        presence_penalty: args.presence_penalty,
+        seed: args.seed,
+        stop: args.stop,
+        stream: args.stream,
+        temperature: args.temperature,
+        top_p: args.top_p,
+        user: args.user,
+        response_format: args.response_format,
+        top_logprobs: args.top_logprobs,
+    }
+
+    return {
+        model: args.model,
+        input: args,
+        modelParams: params
+    }
+}
+
+interface TraceConfig {
+    name?: string
+    inputs?: Record<string, any>
+    config?: WrapperConfig
+}
+
+const generateOutput = (res: any): string => {
+    return 'message' in res.choices[0] ? res.choices[0].message : res.choices[0].text ?? ""
+}
+
+const getUsageDetails = (res: any): Record<string, any> => {
+    return res.usage ?? {}
+}
+
+const processChunks = (chunk: unknown): string => {
+    const _chunk = chunk as OpenAI.ChatCompletionChunk | OpenAI.Completions.Completion
+    if ('delta' in _chunk?.choices[0]) {
+        return _chunk.choices[0].delta?.content || ""
+    } else if ('text' in _chunk?.choices[0]) {
+        return _chunk?.choices[0].text || ""
+    }
+    return ""
+}
+
+class TraceGenerator {
+    name?: string
+    inputs?: Record<string, any>
+    startTime?: Date
+    endTime?: Date
+    outputs?: string
+    client?: Langfuse
+    level?: "DEBUG" | "DEFAULT" | "WARNING" | "ERROR" | undefined
+    trace?: LangfuseTraceClient
+    config?: WrapperConfig
+    constructor(tracerConfig: TraceConfig) {
+        this.config = tracerConfig.config
+        this.name = tracerConfig.name
+        this.inputs = tracerConfig.inputs
+        this.client = client
+        this.startTime = new Date()
+    }
+
+    createTrace(): void {
+        const input = parseInputArgs(this.inputs ?? {})
+        this.trace = client.trace({
+            name: this.config?.trace_name,
+            input: input.input,
+            metadata: this.config?.metadata,
+            tags: this.config?.tags,
+            userId: this.config?.user_id,
+            timestamp: this.startTime,
+            sessionId: this.config?.session_id,
+            version: this.config?.version,
+            release: this.config?.release
+        })
+    }
+
+    createGeneration(
+        output?: string,
+        usage?: Record<string, any>,
```

Comment:
does it make sense to type usage more strictly? promptTokens, completionTokens

Reply:
It does make sense to have typing there. Realised that langfuse doesn't take the openai usage details directly. Have updated in code and added tests for the same :)

Reply:
What kind of translation do you need to do here? generally we tried to take the openai object as it is

Reply:
This is more like a UI test case, which checks if the UI is rendering the chats properly. Backend test doesn't make a lot of sense actually.
Additionally, we can always confirm that the wrapper is not just taking the first element in the array.

Reply:
Removing this test from list.